src/core/mac/mac_header_ie.hpp

```diff
@@ -102,6 +118,32 @@ class HeaderIe
      */
     void SetLength(uint8_t aLength) { mFields.m8[0] = (mFields.m8[0] & ~kLengthMask) | (aLength & kLengthMask); }
 
+    /**
+     * Initializes the Header IE to the given Header IE Content.
+     *
+     * @tparam[in]  ContentType   The type of the Header IE Content.
+     */
+    template <typename ContentType> ContentType *Init(void) { return Init<ContentType>(sizeof(ContentType)); }
+
+    /**
+     * Initializes the Header IE to the given Header IE Content.
+     *
+     * @tparam[in]  ContentType   The type of the Header IE Content.
+     * @param[in]   aLen  The IE content length.
+     */
+    template <typename ContentType> ContentType *Init(uint8_t aLen)
+    {
+        static_assert(std::is_standard_layout<ContentType>::value, "ContentType is not standard layout");
+        Init(ContentType::kHeaderIeId, aLen);
+        return new (this + 1) ContentType();
+    }
+
+    template <typename ContentType> const ContentType *GetContent(void) const
+    {
+        static_assert(std::is_standard_layout<ContentType>::value, "ContentType is not standard layout");
```

Comment:
I am not sure we can use `std:: is_standard_layout ` lib here.
We avoided use of any `std` lib function  to keep OT portable.